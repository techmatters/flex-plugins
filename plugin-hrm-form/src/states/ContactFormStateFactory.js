/**
 * Copyright (C) 2021-2023 Technology Matters
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published
 * by the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see https://www.gnu.org/licenses/.
 */

export const ValidationType = {
  REQUIRED: 'REQUIRED', // Will not be applied if in the callerInformation tab and callType is not caller.  Will not be applied when callType is standalone.
};
export const FieldType = {
  CALL_TYPE: 'CALL_TYPE',
  CHECKBOX: 'CHECKBOX',
  CHECKBOX_FIELD: 'CHECKBOX_FIELD',
  INTERMEDIATE: 'INTERMEDIATE',
  SELECT_SINGLE: 'SELECT_SINGLE',
  TAB: 'TAB',
  TEXT_BOX: 'TEXT_BOX',
  TEXT_INPUT: 'TEXT_INPUT',
};

export function isNotCategory(value) {
  const notCategory = ['error', 'touched', 'type', 'validation'];
  return notCategory.includes(value);
}

export function isNotSubcategory(value) {
  const notSubcategory = ['type'];
  return notSubcategory.includes(value);
}
