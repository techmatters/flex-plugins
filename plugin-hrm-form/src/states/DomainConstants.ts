/**
 * Copyright (C) 2021-2023 Technology Matters
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published
 * by the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see https://www.gnu.org/licenses/.
 */

import { customSmsChannelTypes } from '../utils/smsChannels';

const defaultChannelTypes = {
  voice: 'voice',
  sms: 'sms',
  facebook: 'facebook',
  whatsapp: 'whatsapp',
  web: 'web',
} as const;

const customChannelTypes = {
  telegram: 'telegram',
  instagram: 'instagram',
  line: 'line',
} as const;

/**
 * These are the channels without the custom SMS channels.
 *
 * Since the custom SMS channels are going to behave like the default SMS channel,
 * some components should not be aware of them, like the QueueStatus or
 * the PreviousContactsBanner.
 */
export const coreChannelTypes = {
  ...defaultChannelTypes,
  ...customChannelTypes,
};

/**
 * This is the complete list of channels.
 */
export const channelTypes = {
  ...coreChannelTypes,
  ...customSmsChannelTypes,
};

export type ChannelTypes = typeof channelTypes[keyof typeof channelTypes];
export type CoreChannelTypes = typeof coreChannelTypes[keyof typeof coreChannelTypes];

const chatChannels = [
  channelTypes.whatsapp,
  channelTypes.facebook,
  channelTypes.web,
  channelTypes.modica,
  channelTypes.sms,
  channelTypes.telegram,
  channelTypes.instagram,
  channelTypes.line,
];

export const isVoiceChannel = (channel: string) => channel === channelTypes.voice;
export const isChatChannel = (channel: string) => chatChannels.includes(channel as any);

export type ChannelColors = {
  [C in CoreChannelTypes]: string;
};

export const transferModes = {
  cold: 'COLD',
  warm: 'WARM',
};

export const transferStatuses = {
  transferring: 'transferring',
  accepted: 'accepted',
  rejected: 'rejected',
} as const;
