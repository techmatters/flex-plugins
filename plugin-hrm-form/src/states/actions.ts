import {
  GeneralActionType,
  INITIALIZE_CONTACT_STATE,
  RECREATE_CONTACT_STATE,
  REMOVE_CONTACT_STATE,
  DefinitionVersion,
} from './types';

export const initializeContactState = (definitions: DefinitionVersion) => (taskId: string): GeneralActionType => ({
  type: INITIALIZE_CONTACT_STATE,
  definitions,
  taskId,
});

export const recreateContactState = (definitions: DefinitionVersion) => (taskId: string): GeneralActionType => ({
  type: RECREATE_CONTACT_STATE,
  definitions,
  taskId,
});

export const removeContactState = (taskId: string): GeneralActionType => ({
  type: REMOVE_CONTACT_STATE,
  taskId,
});
