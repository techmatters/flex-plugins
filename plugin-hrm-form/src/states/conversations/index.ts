/**
 * Copyright (C) 2021-2023 Technology Matters
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published
 * by the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see https://www.gnu.org/licenses/.
 */

/**
 * Currently this state is only used when the enable_aselo_messaging_ui feature flag is set.
 * Otherwise internal Twilio flex state us used along with the default twilio message UI
 */

import { createAction, createAsyncAction, createReducer } from 'redux-promise-middleware-actions';
import { Conversation } from '@twilio/conversations';
import { ConversationsState } from '@twilio/flex-ui';

export enum MessageSendStatus {
  SENDING = 'SENDING',
  NOT_SENDING = 'NOT_SENDING',
}

type ConversationsState = {
  [conversationSid: string]: {
    draftMessageText: string;
    messageSendStatus: MessageSendStatus;
  };
};

const initialState: ConversationsState = {};

const UPDATE_DRAFT_MESSAGE_TEXT_ACTION = 'UPDATE_DRAFT_MESSAGE_TEXT';

export const newUpdateDraftMessageTextAction = createAction(
  UPDATE_DRAFT_MESSAGE_TEXT_ACTION,
  (conversationSid: string, draftMessageText: string) => ({ conversationSid, draftMessageText }),
);

const SEND_MESSAGE_ACTION = 'SEND_MESSAGE_ACTION';

export const newSendMessageeAsyncAction = createAsyncAction(
  SEND_MESSAGE_ACTION,
  async (conversation: Conversation, messageText: string): Promise<{ returnStatus: number; sid: string }> => {
    const returnStatus = await conversation.sendMessage(messageText);
    return { returnStatus, sid: conversation.sid };
  },
  (conversation: Conversation, messageText: string) => ({ conversation, messageText }),
);

export const reduce = createReducer(initialState, handleAction => [
  handleAction(newUpdateDraftMessageTextAction, (state, action) => {
    return {
      ...state,
      [action.payload.conversationSid]: {
        draftMessageText: action.payload.draftMessageText,
        messageSendStatus: MessageSendStatus.NOT_SENDING,
      },
    };
  }),
  handleAction(newSendMessageeAsyncAction.pending as typeof newSendMessageeAsyncAction, (state, action) => {
    return {
      ...state,
      [action.meta.conversation.sid]: {
        draftMessageText: action.meta.messageText,
        messageSendStatus: MessageSendStatus.SENDING,
      },
    };
  }),
  handleAction(newSendMessageeAsyncAction.fulfilled, (state, action) => {
    return {
      ...state,
      [action.payload.sid]: {
        draftMessageText: '',
        messageSendStatus: MessageSendStatus.NOT_SENDING,
      },
    };
  }),
  handleAction(newSendMessageeAsyncAction.rejected, (state, action) => {
    const meta = (action as any).meta as ReturnType<typeof newSendMessageeAsyncAction>['meta'];
    return {
      ...state,
      [meta.conversation.sid]: {
        draftMessageText: '',
        messageSendStatus: MessageSendStatus.NOT_SENDING,
      },
    };
  }),
]);
