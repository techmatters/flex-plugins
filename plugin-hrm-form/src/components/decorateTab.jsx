/**
 * Copyright (C) 2021-2023 Technology Matters
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published
 * by the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see https://www.gnu.org/licenses/.
 */

import React from 'react';
import ErrorIcon from '@material-ui/icons/Error';
import { Template } from '@twilio/flex-ui';

import { StyledTab } from '../styles/HrmStyles';
import { formIsValid } from '../states/ValidationRules';

const decorateTab = (label, formRoot) => {
  if (formIsValid(formRoot)) {
    return <StyledTab key={label} label={<Template code={label} />} />;
  }
  return (
    <StyledTab
      key={label}
      label={
        <>
          <div style={{ verticalAlign: 'middle' }}>
            <ErrorIcon fontSize="inherit" style={{ color: 'red' }} />
          </div>{' '}
          <Template code={label} />
        </>
      }
    />
  );
};

decorateTab.displayName = 'decorateTab';

export default decorateTab;
