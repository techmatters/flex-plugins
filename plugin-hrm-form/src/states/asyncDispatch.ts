/**
 * Copyright (C) 2021-2023 Technology Matters
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published
 * by the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see https://www.gnu.org/licenses/.
 */

import { Dispatch } from 'react';
import { AnyAction } from 'redux-promise-middleware-actions/lib/actions';

/**
 * Wraps a regular dispatcher in a try ... catch so that rejections are only routed to the '_REJECTED' action handler.
 * This is rather than also being passed up the stack as an unhandled exception, which is the default behaviour of react-promise-middleware
 * @param innerDispatch - the dispatch function to wrap
 */
const asyncDispatch = <T extends AnyAction>(innerDispatch: Dispatch<T>) => async (action: T) => {
  try {
    await innerDispatch(action);
  } catch (err) {
    console.debug(`Rejected promise being handled by ${action.type}_REJECTED:`, err);
  }
};

export default asyncDispatch;
