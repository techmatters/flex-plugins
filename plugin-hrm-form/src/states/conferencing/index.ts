/**
 * Copyright (C) 2021-2023 Technology Matters
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published
 * by the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see https://www.gnu.org/licenses/.
 */
import { createAction, createReducer } from 'redux-promise-middleware-actions';
import { omit } from 'lodash';

import { GeneralActionType, INITIALIZE_CONTACT_STATE, RECREATE_CONTACT_STATE } from '../types';
import { removeContactState } from '../actions';
import { RootState, conferencingBase, namespace } from '..';

export type CallStatus =
  | 'no-call'
  | 'initiating'
  | 'initiated'
  | 'ringing'
  | 'busy'
  | 'failed'
  | 'in-progress'
  | 'completed';

export type ConferencingState = {
  tasks: {
    [taskId: string]: {
      isDialogOpen: boolean;
      callStatus: CallStatus;
      phoneNumber: string;
    };
  };
};

export const newTaskEntry = {
  isDialogOpen: false,
  callStatus: 'no-call' as CallStatus,
  phoneNumber: '',
};

const SET_IS_DIALOG_OPEN = 'conferencing/set-is-dialog-open';

export const setIsDialogOpenAction = createAction(SET_IS_DIALOG_OPEN, (taskId: string, isDialogOpen: boolean) => ({
  taskId,
  isDialogOpen,
}));

const SET_CALL_STATUS = 'conferencing/set-call-status';

export const setCallStatusAction = createAction(SET_CALL_STATUS, (taskId: string, callStatus: CallStatus) => ({
  taskId,
  callStatus,
}));

const SET_PHONE_NUMBER = 'conferencing/set-phone-number';

export const setPhoneNumberAction = createAction(SET_PHONE_NUMBER, (taskId: string, phoneNumber: string) => ({
  taskId,
  phoneNumber,
}));

type ConferencingStateAction =
  | ReturnType<typeof setIsDialogOpenAction>
  | ReturnType<typeof setCallStatusAction>
  | ReturnType<typeof setPhoneNumberAction>;

const initialState: ConferencingState = {
  tasks: {},
};

const createNewEntryForTaskId = (state: ConferencingState, payload: GeneralActionType) => {
  return {
    ...state,
    tasks: {
      ...state.tasks,
      [payload.taskId]: newTaskEntry,
    },
  };
};

const conferencingReducer = createReducer(initialState, handleAction => [
  // Handle GeneralActionType
  {
    [INITIALIZE_CONTACT_STATE]: createNewEntryForTaskId,
  },
  {
    [RECREATE_CONTACT_STATE]: createNewEntryForTaskId,
  },
  handleAction(removeContactState, (state, { taskId }) => {
    if (state.tasks[taskId]) {
      return {
        ...state,
        tasks: omit(state.tasks, taskId),
      };
    }

    return state;
  }),

  // Handle ConferencingStateAction
  handleAction(setIsDialogOpenAction, (state, { payload }) => ({
    ...state,
    tasks: {
      ...state.tasks,
      [payload.taskId]: {
        ...state.tasks[payload.taskId],
        isDialogOpen: payload.isDialogOpen,
      },
    },
  })),
  handleAction(setCallStatusAction, (state, { payload }) => ({
    ...state,
    tasks: {
      ...state.tasks,
      [payload.taskId]: {
        ...state.tasks[payload.taskId],
        callStatus: payload.callStatus,
      },
    },
  })),
  handleAction(setPhoneNumberAction, (state, { payload }) => ({
    ...state,
    tasks: {
      ...state.tasks,
      [payload.taskId]: {
        ...state.tasks[payload.taskId],
        phoneNumber: payload.phoneNumber,
      },
    },
  })),
]);

// eslint-disable-next-line import/no-unused-modules
export const reduce = (
  inputState = initialState,
  action: ConferencingStateAction | GeneralActionType,
): ConferencingState => {
  return conferencingReducer(inputState, action);
};

export const isCallStatusLoading = (callStatus: CallStatus) =>
  callStatus === 'initiating' || callStatus === 'initiated' || callStatus === 'ringing';
