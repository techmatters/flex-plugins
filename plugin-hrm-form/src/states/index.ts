/**
 * Copyright (C) 2021-2023 Technology Matters
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published
 * by the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see https://www.gnu.org/licenses/.
 */

import type { FlexState } from '@twilio/flex-ui';
import { combineReducers } from 'redux';

import { reduce as ContactStateReducer } from './contacts/reducer';
import { reduce as SearchFormReducer } from './search/reducer';
import { reduce as ConnectedCaseReducer } from './case/reducer';
import { reduce as CaseListReducer } from './caseList/reducer';
import { reduce as QueuesStatusReducer } from './queuesStatus/reducer';
import { reduce as ConfigurationReducer } from './configuration/reducer';
import { reduce as RoutingReducer } from './routing/reducer';
import { reduce as CSAMReportReducer } from './csam-report/reducer';
import { reduce as DualWriteReducer } from './dualWrite/reducer';
import { reduce as ReferrableResourcesReducer } from './resources';
import { reduce as ConversationsReducer } from './conversations';
import { reduce as ConferencingReducer } from './conferencing';
import { CaseState } from './case/types';

// Register your redux store under a unique namespace
export const namespace = 'plugin-hrm-form';
export const contactFormsBase = 'activeContacts';
export const searchContactsBase = 'searchContacts';
export const caseListBase = 'caseList';
export const connectedCaseBase = 'connectedCase';
export const queuesStatusBase = 'queuesStatusState';
export const configurationBase = 'configuration';
export const conversationsBase = 'conversations';
export const routingBase = 'routing';
export const csamReportBase = 'csam-report';
export const dualWriteBase = 'dualWrite';
export const referrableResourcesBase = 'referrableResources';
// eslint-disable-next-line import/no-unused-modules
export const conferencingBase = 'conferencing';

const reducers = {
  [contactFormsBase]: ContactStateReducer,
  [searchContactsBase]: SearchFormReducer,
  [queuesStatusBase]: QueuesStatusReducer,
  [caseListBase]: CaseListReducer,
  [configurationBase]: ConfigurationReducer,
  [routingBase]: RoutingReducer,
  [csamReportBase]: CSAMReportReducer,
  [dualWriteBase]: DualWriteReducer,
  [referrableResourcesBase]: ReferrableResourcesReducer,
  [conversationsBase]: ConversationsReducer,
  [conferencingBase]: ConferencingReducer,
  /*
   * [csamClcReportBase]: CSAMCLCReportReducer,
   * [connectedCaseBase] - this is going to be combined manually, rather than using 'combineReducers', so isn't in this map
   */
};
type HrmState = {
  [P in keyof typeof reducers]: ReturnType<typeof reducers[P]>;
} & { [connectedCaseBase]: CaseState };

export type RootState = FlexState & { [namespace]: HrmState };
const combinedReducers = combineReducers(reducers);

// Combine the reducers
const reducer = (state: HrmState, action): HrmState => {
  return {
    ...combinedReducers(state, action),
    /*
     * ConnectedCaseReducer's signature includes a parameter for global Hrm State as well as the specific CaseState
     * This makes it incompatible with combineReducers, so instead, we add the case state property with an explicit call to ConnectedCaseReducer, where we specify the extra parameter
     */
    [connectedCaseBase]: ConnectedCaseReducer(state, (state ?? {})[connectedCaseBase], action),
  };
};

export default reducer;
