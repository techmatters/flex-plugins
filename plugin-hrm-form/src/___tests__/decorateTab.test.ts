/**
 * Copyright (C) 2021-2023 Technology Matters
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published
 * by the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see https://www.gnu.org/licenses/.
 */

import renderer from 'react-test-renderer';

import { formIsValid } from '../states/validationRules';
import decorateTab from '../components/decorateTab';

jest.mock('../styles/HrmStyles', () => ({ StyledTab: 'StyledTab' }));
jest.mock('../states/validationRules', () => ({ formIsValid: jest.fn() }));
jest.mock('@material-ui/icons/Error', () => 'ErrorIcon');

const mockFormIsValid = formIsValid as jest.Mock<ReturnType<typeof formIsValid>>;

test('decorateTab when valid', () => {
  mockFormIsValid.mockReturnValueOnce(true);
  const component = renderer.create(decorateTab('My Label', {}));
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('decorateTab when invalid', () => {
  mockFormIsValid.mockReturnValueOnce(false);
  const component = renderer.create(decorateTab('My Label', {}));
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
