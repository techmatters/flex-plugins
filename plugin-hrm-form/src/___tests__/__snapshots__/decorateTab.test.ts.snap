// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`decorateTab when invalid 1`] = `
<StyledTab
  label={
    <React.Fragment>
      <div
        style={
          Object {
            "verticalAlign": "middle",
          }
        }
      >
        <ErrorIcon
          fontSize="inherit"
          style={
            Object {
              "color": "red",
            }
          }
        />
      </div>
       
      <t
        className="Twilio"
        code="My Label"
      />
    </React.Fragment>
  }
/>
`;

exports[`decorateTab when valid 1`] = `
<StyledTab
  label={
    <t
      className="Twilio"
      code="My Label"
    />
  }
/>
`;
